network/dag/consistency.go

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright (C) 2023 Nuts community
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package dag
+
+import (
+	"context"
+	"github.com/nuts-foundation/go-stoabs"
+	"github.com/nuts-foundation/nuts-node/network/dag/tree"
+	"github.com/nuts-foundation/nuts-node/network/log"
+	"sync"
+	"time"
+)
+
+type circuit int
+
+const (
+	circuitGreen circuit = iota
+	circuitYellow
+	circuitRed
+)
+
+// xorTreeRepair is responsible for repairing the XOR tree. Its loop is triggered when the network layer detects differences in XOR values with all other nodes.
+// It will loop over all pages of the XOR tree and recalculates the XOR value with the transactions in the database.
+// This repair is needed because current networks have nodes that have a wrong XOR value. How this happens is not yet known, it could be due to DB failures of due to failures in older versions.
+// The fact is that we can fix the state relatively easy.
+// The loop checks a page (512 LC values) per 10 seconds and continues looping until the network layer signals all is ok again.
+type xorTreeRepair struct {
+	ctx          context.Context
+	cancel       context.CancelFunc
+	ticker       *time.Ticker
+	currentPage  uint32
+	state        *state
+	circuitState circuit
+	mutex        sync.Mutex
+}
+
+func newXorTreeRepair(state *state) *xorTreeRepair {
+	return &xorTreeRepair{
+		state:  state,
+		ticker: time.NewTicker(10 * time.Second),
+	}
+}
+
+func (f *xorTreeRepair) start() {
+	var ctx context.Context
+	ctx, f.cancel = context.WithCancel(context.Background())
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-f.ticker.C:
+				f.checkPage()
+			}
+		}
+	}()
+}
+
+func (f *xorTreeRepair) shutdown() {
+	if f.cancel != nil {
+		f.cancel()
+	}
+}
+
+func (f *xorTreeRepair) checkPage() {
+	f.mutex.Lock()
+	defer f.mutex.Unlock()
+
+	// ignore run if circuit is not red
+	if f.circuitState < circuitRed {
+		return
+	}
+
+	currentLC := f.state.lamportClockHigh.Load()
+	lcStart := f.currentPage * PageSize
+	lcEnd := lcStart + PageSize
+
+	// initialize an XOR tree
+	calculatedXorTree := tree.New(tree.NewXor(), PageSize)
+
+	// acquire global lock
+	err := f.state.graph.db.Write(context.Background(), func(txn stoabs.WriteTx) error {
```

Comment:
how quick/fast is this? The funcs to be called by externals are now blocked by the mutex.(although they are blocked by the write TX anyways..?)

Reply:
takes around 50-100ms per check.

Reply:
perhaps good to do a performance test with this? 100ms on an interval of a page/10s = ~1% DB downtime. It could have a significant impact during bursts of data (auto publishing OpenID services), but becomes less relevant as more transactions move off the dag.